File: scripts/profile-pdf.py
#!/usr/bin/env python3

import base64
import html
import io
import json
import os
import sys
from datetime import datetime

from reportlab.lib import colors
from reportlab.lib.enums import TA_LEFT
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import mm
from reportlab.lib.utils import ImageReader
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from reportlab.platypus import Flowable, KeepTogether, Paragraph, SimpleDocTemplate, Spacer, Table, TableStyle


GREEN = colors.HexColor("#197A52")
GREEN_DARK = colors.HexColor("#0F4F36")
GREEN_SOFT = colors.HexColor("#EDF7F1")
NAVY = colors.HexColor("#112A23")
INK = colors.HexColor("#17231E")
MUTED = colors.HexColor("#66736D")
BORDER = colors.HexColor("#D9E6DF")
SURFACE = colors.HexColor("#F8FBF9")
AMBER = colors.HexColor("#B56D12")
LOGO_PATH = ""


def register_fonts():
    candidates = [
        "/System/Library/Fonts/Supplemental/Arial Unicode.ttf",
        "/System/Library/Fonts/Supplemental/Arial.ttf",
        "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
    ]
    for path in candidates:
        if os.path.exists(path):
            pdfmetrics.registerFont(TTFont("AmirosSans", path))
            return "AmirosSans"
    return "Helvetica"


FONT = register_fonts()


def clean_text(value):
    return (str(value or "").replace("\u2010", "-").replace("\u2011", "-")
            .replace("\u2012", "-").replace("\u2013", "-").replace("\u2014", "-")
            .replace("\u2212", "-").strip())


def safe(value):
    return html.escape(clean_text(value)).replace("\n", "<br/>")


def format_date(value):
    try:
        return datetime.fromisoformat(str(value).replace("Z", "+00:00")).strftime("%B %d, %Y at %H:%M")
    except Exception:
        return clean_text(value)


class ProfileAvatar(Flowable):
    def __init__(self, name, image_payload=None, size=28 * mm):
        super().__init__()
        self.width = self.height = size
        self.name = clean_text(name)
        self.reader = None
        try:
            encoded = (image_payload or {}).get("data")
            if encoded:
                self.reader = ImageReader(io.BytesIO(base64.b64decode(encoded)))
        except Exception:
            self.reader = None

    def draw(self):
        radius = self.width / 2
        self.canv.saveState()
        path = self.canv.beginPath()
        path.circle(radius, radius, radius)
        self.canv.clipPath(path, stroke=0, fill=0)
        if self.reader:
            source_width, source_height = self.reader.getSize()
            scale = max(self.width / source_width, self.height / source_height)
            draw_width, draw_height = source_width * scale, source_height * scale
            self.canv.drawImage(self.reader, (self.width - draw_width) / 2, (self.height - draw_height) / 2,
                                width=draw_width, height=draw_height, mask="auto")
        else:
            self.canv.setFillColor(GREEN_SOFT)
            self.canv.rect(0, 0, self.width, self.height, fill=1, stroke=0)
            initials = "".join(part[:1] for part in self.name.split()[:2]).upper() or "A"
            self.canv.setFillColor(GREEN_DARK)
            self.canv.setFont(FONT, 13)
            self.canv.drawCentredString(radius, radius - 4, initials)
        self.canv.restoreState()
        self.canv.setStrokeColor(BORDER)
        self.canv.setLineWidth(0.8)
        self.canv.circle(radius, radius, radius, fill=0, stroke=1)


def draw_page(canvas, document):
    width, height = A4
    canvas.saveState()
    if LOGO_PATH and os.path.exists(LOGO_PATH):
        canvas.drawImage(LOGO_PATH, 18 * mm, height - 27 * mm, width=18 * mm, height=18 * mm,
                         preserveAspectRatio=True, anchor="c", mask="auto")
    canvas.setFillColor(INK)
    canvas.setFont(FONT, 16)
    canvas.drawString(35 * mm, height - 18.5 * mm, "AmirOS")
    canvas.setFillColor(MUTED)
    canvas.setFont(FONT, 8)
    canvas.drawString(35 * mm, height - 23 * mm, "Private relationship intelligence")
    canvas.setStrokeColor(BORDER)
    canvas.line(18 * mm, height - 30 * mm, width - 18 * mm, height - 30 * mm)
    canvas.setFillColor(MUTED)
    canvas.setFont(FONT, 7.5)
    canvas.drawString(18 * mm, 12 * mm, "Private analysis generated locally by AmirOS")
    canvas.drawRightString(width - 18 * mm, 12 * mm, f"Page {document.page}")
    canvas.restoreState()


def build_pdf(payload):
    global LOGO_PATH
    LOGO_PATH = clean_text(payload.get("logoPath"))
    output = io.BytesIO()
    document = SimpleDocTemplate(
        output, pagesize=A4, rightMargin=18 * mm, leftMargin=18 * mm,
        topMargin=37 * mm, bottomMargin=20 * mm,
        title=f"AmirOS profile - {clean_text(payload.get('contactName'))}",
        author="AmirOS", subject="Private relationship profile analysis",
    )
    styles = getSampleStyleSheet()
    body = ParagraphStyle("Body", parent=styles["BodyText"], fontName=FONT, fontSize=9.6,
                          leading=14.2, textColor=INK, spaceAfter=2.2 * mm)
    title = ParagraphStyle("ProfileTitle", parent=body, fontSize=24, leading=28,
                           textColor=INK, spaceAfter=1.4 * mm)
    subtitle = ParagraphStyle("Subtitle", parent=body, fontSize=9, leading=13, textColor=MUTED)
    eyebrow = ParagraphStyle("Eyebrow", parent=body, fontSize=7.7, leading=9, textColor=GREEN,
                             spaceAfter=1.5 * mm)
    section = ParagraphStyle("Section", parent=body, fontSize=13, leading=17, textColor=GREEN_DARK,
                             spaceBefore=5 * mm, spaceAfter=2.5 * mm)
    subsection = ParagraphStyle("Subsection", parent=body, fontSize=10.5, leading=14,
                                textColor=NAVY, spaceBefore=2.5 * mm, spaceAfter=1.5 * mm)
    bullet = ParagraphStyle("Bullet", parent=body, leftIndent=4 * mm, firstLineIndent=-3 * mm,
                            bulletIndent=0, spaceAfter=1.3 * mm)
    small = ParagraphStyle("Small", parent=body, fontSize=8, leading=11.5, textColor=MUTED)
    metric = ParagraphStyle("Metric", parent=body, fontSize=15, leading=18, textColor=GREEN_DARK)

    contact_name = clean_text(payload.get("contactName")) or "WhatsApp contact"
    profile_kind = "GROUP RELATIONSHIP PROFILE" if payload.get("isGroup") else "PRIVATE CONTACT PROFILE"
    hero_copy = [
        Paragraph(profile_kind, eyebrow),
        Paragraph(safe(contact_name), title),
        Paragraph(f"{safe(payload.get('relationship') or 'Contact')} | {safe(payload.get('tone') or 'Automatic')} tone | {safe(payload.get('language') or 'Automatic')}", subtitle),
    ]
    hero = Table([[ProfileAvatar(contact_name, payload.get("profileImage")), hero_copy]], colWidths=[35 * mm, 127 * mm])
    hero.setStyle(TableStyle([
        ("BACKGROUND", (0, 0), (-1, -1), SURFACE), ("BOX", (0, 0), (-1, -1), 0.7, BORDER),
        ("VALIGN", (0, 0), (-1, -1), "MIDDLE"), ("LEFTPADDING", (0, 0), (0, 0), 8),
        ("RIGHTPADDING", (0, 0), (0, 0), 8), ("LEFTPADDING", (1, 0), (1, 0), 12),
        ("RIGHTPADDING", (1, 0), (1, 0), 12), ("TOPPADDING", (0, 0), (-1, -1), 11),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 11),
    ]))

    source_count = int(payload.get("sourceMessageCount") or 0)
    manual_items = payload.get("manualMemory") or []
    insights = payload.get("insights") or []
    commitments = payload.get("commitments") or []
    metrics = Table([[Paragraph(f"<b>{source_count}</b><br/><font size='8' color='#66736D'>tracked messages</font>", metric),
                      Paragraph(f"<b>{len(insights)}</b><br/><font size='8' color='#66736D'>active insights</font>", metric),
                      Paragraph(f"<b>{len(commitments)}</b><br/><font size='8' color='#66736D'>open commitments</font>", metric)]],
                    colWidths=[54 * mm, 54 * mm, 54 * mm])
    metrics.setStyle(TableStyle([
        ("BACKGROUND", (0, 0), (-1, -1), GREEN_SOFT), ("BOX", (0, 0), (-1, -1), 0.6, BORDER),
        ("INNERGRID", (0, 0), (-1, -1), 0.4, BORDER), ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
        ("ALIGN", (0, 0), (-1, -1), "CENTER"), ("TOPPADDING", (0, 0), (-1, -1), 9),
        ("BOTTOMPADDING", (0, 0), (-1, -1), 9),
    ]))

    story = [hero, Spacer(1, 4 * mm), metrics, Paragraph("Profile analysis", section)]
    for raw_line in clean_text(payload.get("summary")).splitlines():
        line = raw_line.strip()
        if not line:
            story.append(Spacer(1, 1.2 * mm))
        elif line.startswith(("-", "•", "*")):
            story.append(Paragraph(f"- {safe(line[1:].strip())}", bullet))
        elif len(line) <= 75 and not line.endswith((".", "!", "?", ":")):
            story.append(Paragraph(f"<b>{safe(line)}</b>", subsection))
        else:
            story.append(Paragraph(safe(line), body))

    if insights:
        rows = [[Paragraph("Signal", small), Paragraph("Evidence-backed detail", small), Paragraph("Confidence", small)]]
        for item in insights[:16]:
            evidence = item.get("evidence") or {}
            detail = f"<b>{safe(item.get('content'))}</b><br/><font size='7.5' color='#66736D'>{safe(evidence.get('senderName') or contact_name)}: {safe(evidence.get('excerpt'))}</font>"
            rows.append([Paragraph(safe(item.get("kind", "fact")).replace("_", " ").title(), small), Paragraph(detail, body), Paragraph(f"{round(float(item.get('confidence') or 0) * 100)}%", small)])
        table = Table(rows, colWidths=[28 * mm, 112 * mm, 22 * mm], repeatRows=1)
        table.setStyle(TableStyle([
            ("BACKGROUND", (0, 0), (-1, 0), NAVY), ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
            ("GRID", (0, 0), (-1, -1), 0.45, BORDER), ("VALIGN", (0, 0), (-1, -1), "TOP"),
            ("LEFTPADDING", (0, 0), (-1, -1), 7), ("RIGHTPADDING", (0, 0), (-1, -1), 7),
            ("TOPPADDING", (0, 0), (-1, -1), 6), ("BOTTOMPADDING", (0, 0), (-1, -1), 6),
            ("ROWBACKGROUNDS", (0, 1), (-1, -1), [colors.white, SURFACE]),
        ]))
        story.extend([Paragraph("Relationship intelligence", section), table])

    if commitments:
        block = [Paragraph("Open commitments", section)]
        for item in commitments:
            owner = "Amir" if item.get("owner") == "me" else clean_text(item.get("assigneeName")) or "Contact"
            block.append(Paragraph(f"- <b>{safe(owner)}</b>: {safe(item.get('content'))}", bullet))
        story.append(KeepTogether(block))

    style_profile = payload.get("styleProfile") or {}
    if style_profile:
        style_rows = [[Paragraph("Message length", small), Paragraph("Emoji use", small), Paragraph("Formality", small)],
                      [Paragraph(safe(style_profile.get("messageLength")), body), Paragraph(safe(style_profile.get("emojiUse")), body), Paragraph(safe(style_profile.get("formality")), body)]]
        style_table = Table(style_rows, colWidths=[54 * mm, 54 * mm, 54 * mm])
        style_table.setStyle(TableStyle([("BACKGROUND", (0, 0), (-1, -1), SURFACE), ("GRID", (0, 0), (-1, -1), 0.5, BORDER),
                                         ("VALIGN", (0, 0), (-1, -1), "TOP"), ("LEFTPADDING", (0, 0), (-1, -1), 8),
                                         ("RIGHTPADDING", (0, 0), (-1, -1), 8), ("TOPPADDING", (0, 0), (-1, -1), 7),
                                         ("BOTTOMPADDING", (0, 0), (-1, -1), 7)]))
        style_block = [Paragraph("Communication style", section), Paragraph(safe(style_profile.get("summary")), body), style_table]
        style_block.extend(Paragraph(f"- {safe(guidance)}", bullet) for guidance in style_profile.get("replyGuidance") or [])
        story.append(KeepTogether(style_block))

    group_summary = payload.get("groupSummary") or {}
    if group_summary:
        story.extend([Paragraph("Group dynamics", section), Paragraph(safe(group_summary.get("summary")), body)])
        for heading, key in [("Decisions", "decisions"), ("Tasks", "tasks"), ("Open questions", "unansweredQuestions")]:
            values = group_summary.get(key) or []
            if values:
                story.append(Paragraph(f"<b>{heading}</b>", subsection))
                story.extend(Paragraph(f"- {safe(item)}", bullet) for item in values)

    if manual_items:
        memory_content = [Paragraph("Operator-saved memory", section)]
        memory_content.extend(Paragraph(f"- {safe(item)}", bullet) for item in manual_items)
        story.append(KeepTogether(memory_content))

    generated_at = format_date(payload.get("generatedAt"))
    note = (f"<b>Analysis details</b><br/>Generated {safe(generated_at)} from {source_count} tracked incoming messages "
            f"and {len(manual_items)} manual memory items. Inferences are evidence-backed where available, but may still be incomplete and should be reviewed.")
    story.extend([Spacer(1, 5 * mm), Table([[Paragraph(note, small)]], colWidths=[162 * mm], style=TableStyle([
        ("BACKGROUND", (0, 0), (-1, -1), SURFACE), ("BOX", (0, 0), (-1, -1), 0.6, BORDER),
        ("LEFTPADDING", (0, 0), (-1, -1), 11), ("RIGHTPADDING", (0, 0), (-1, -1), 11),
        ("TOPPADDING", (0, 0), (-1, -1), 9), ("BOTTOMPADDING", (0, 0), (-1, -1), 9),
    ]))])

    document.build(story, onFirstPage=draw_page, onLaterPages=draw_page)
    return output.getvalue()


def main():
    payload = json.load(sys.stdin)
    sys.stdout.buffer.write(build_pdf(payload))


if __name__ == "__main__":
    main()

File: src/profile-pdf.ts
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { createRequire } from "node:module";
import { dirname, resolve } from "node:path";
import { pathToFileURL } from "node:url";
import type {
  ContactInsight,
  ContactMemoryItem,
  ContactPreferences,
  ContactProfile,
  GroupConversationSummary,
  RelationshipCommitment,
  WritingStyleProfile,
} from "./amiros-state.js";

export type ContactProfilePdfInput = {
  contactName: string;
  contact: ContactPreferences;
  profile: ContactProfile;
  manualMemory: ContactMemoryItem[];
  isGroup: boolean;
  profileImage?: { data: string; mimetype: string };
  insights: ContactInsight[];
  commitments: RelationshipCommitment[];
  styleProfile?: WritingStyleProfile;
  groupSummary?: GroupConversationSummary;
  generatedAt?: number;
  timezoneOffsetMinutes?: number;
  locale?: string;
};

function cleanText(value: unknown): string {
  return String(value || "")
    .replace(/[\u2010\u2011\u2012\u2013\u2014\u2212]/g, "-")
    .trim();
}

function escapeHtml(value: unknown): string {
  return cleanText(value).replace(/[&<>"']/g, (character) => ({
    "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;",
  })[character] || character);
}

function text(value: unknown, tag = "span", className = ""): string {
  return `<${tag}${className ? ` class="${className}"` : ""} dir="auto">${escapeHtml(value)}</${tag}>`;
}

function formatDeviceTime(timestamp: number, offsetMinutes: number, locale: string): string {
  const shifted = new Date(timestamp - offsetMinutes * 60_000);
  try {
    return new Intl.DateTimeFormat(locale, {
      dateStyle: "long", timeStyle: "short", timeZone: "UTC",
    }).format(shifted);
  } catch {
    return new Intl.DateTimeFormat("en", {
      dateStyle: "long", timeStyle: "short", timeZone: "UTC",
    }).format(shifted);
  }
}

function summaryHtml(summary: string): string {
  const blocks: string[] = [];
  let list: string[] = [];
  const flushList = () => {
    if (!list.length) return;
    blocks.push(`<ul>${list.map((item) => `<li dir="auto">${escapeHtml(item)}</li>`).join("")}</ul>`);
    list = [];
  };
  for (const raw of cleanText(summary).split(/\r?\n/)) {
    const line = raw.trim();
    if (!line) { flushList(); continue; }
    if (/^[-*•]\s*/u.test(line)) { list.push(line.replace(/^[-*•]\s*/u, "")); continue; }
    flushList();
    if (line.length <= 80 && !/[.!?:]$/u.test(line)) blocks.push(text(line, "h3"));
    else blocks.push(text(line, "p"));
  }
  flushList();
  return blocks.join("");
}

function chromeExecutable(): string {
  const candidates = [
    process.env.PUPPETEER_EXECUTABLE_PATH?.trim(),
    process.env.CHROME_BIN?.trim(),
    "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
    "/Applications/Chromium.app/Contents/MacOS/Chromium",
    "/usr/bin/google-chrome",
    "/usr/bin/chromium",
  ].filter((candidate): candidate is string => Boolean(candidate));
  const executable = candidates.find(existsSync);
  if (!executable) throw new Error("Chrome is required to export Unicode profile PDFs");
  return executable;
}

async function imageDataUrl(path: string): Promise<string> {
  const data = await readFile(path);
  return `data:image/png;base64,${data.toString("base64")}`;
}

function renderPdfHtml(input: ContactProfilePdfInput, logoUrl: string): string {
  const generatedAt = input.generatedAt || Date.now();
  const offset = Number.isFinite(input.timezoneOffsetMinutes)
    ? input.timezoneOffsetMinutes as number
    : new Date().getTimezoneOffset();
  const locale = input.locale?.trim().slice(0, 35) || "en";
  const generatedLabel = formatDeviceTime(generatedAt, offset, locale);
  const contactName = cleanText(input.contactName) || "WhatsApp contact";
  const profileImageUrl = input.profileImage
    ? `data:${input.profileImage.mimetype};base64,${input.profileImage.data}`
    : "";
  const initials = contactName.split(/\s+/).slice(0, 2).map((part) => part[0] || "").join("").toUpperCase() || "A";
  const insights = input.insights.filter((item) => item.status !== "outdated").slice(-30);
  const commitments = input.commitments.filter((item) => item.status === "open").slice(-20);
  const insightHtml = insights.map((item) => `<article class="signal">
    <div class="signal-heading">${text(item.kind.replaceAll("_", " "), "strong")}<span>${Math.round(item.confidence * 100)}%</span></div>
    ${text(item.content, "p")}
    <blockquote dir="auto">${escapeHtml(item.evidence.senderName || contactName)}: ${escapeHtml(item.evidence.excerpt)}</blockquote>
  </article>`).join("");
  const commitmentHtml = commitments.map((item) => {
    const owner = item.owner === "me" ? "Amir" : item.assigneeName || "Contact";
    return `<li class="commitment"><strong dir="auto">${escapeHtml(owner)}</strong><span dir="auto">${escapeHtml(item.content)}</span></li>`;
  }).join("");
  const style = input.styleProfile;
  const group = input.groupSummary;
  const memoryHtml = input.manualMemory.map((item) => `<li dir="auto">${escapeHtml(item.content)}</li>`).join("");

  return `<!doctype html><html lang="${escapeHtml(locale)}"><head><meta charset="utf-8"><title>AmirOS profile — ${escapeHtml(contactName)}</title><style>
    @page { size: A4; margin: 10mm 15mm; }
    * { box-sizing: border-box; }
    html { color: #17231e; background: white; font-family: -apple-system, BlinkMacSystemFont, "SF Pro Text", Arial, "Arial Hebrew", "Apple Color Emoji", "Segoe UI Emoji", sans-serif; }
    body { margin: 0; font-size: 9.4pt; line-height: 1.43; }
    [dir="rtl"] { text-align: right; unicode-bidi: plaintext; }
    header { display: flex; align-items: center; height: 10mm; margin-bottom: 4mm; border-bottom: .35mm solid #d9e6df; }
    header img { width: 9mm; height: 9mm; object-fit: contain; margin-right: 2.5mm; }
    header strong { display: block; color: #112a23; font-size: 13pt; line-height: 1.05; }
    header small { color: #66736d; font-size: 7.5pt; }
    footer { display: flex; justify-content: space-between; margin-top: 4mm; padding-top: 2.5mm; border-top: .3mm solid #d9e6df; color: #66736d; font-size: 7.2pt; break-inside: avoid; }
    .hero { display: grid; grid-template-columns: 24mm 1fr; gap: 5mm; align-items: center; min-height: 32mm; padding: 4.5mm 6mm; border: .35mm solid #d9e6df; border-radius: 4mm; background: linear-gradient(135deg,#f8fbf9,#edf7f1); break-inside: avoid; }
    .avatar { display: grid; place-items: center; width: 22mm; height: 22mm; overflow: hidden; border: .8mm solid white; border-radius: 50%; color: #0f4f36; background: #dcefe4; box-shadow: 0 1.5mm 4mm rgba(17,42,35,.12); font-size: 14pt; font-weight: 800; }
    .avatar img { width: 100%; height: 100%; object-fit: cover; }
    .eyebrow { margin: 0 0 1mm; color: #197a52; font-size: 7pt; font-weight: 800; letter-spacing: .09em; text-transform: uppercase; }
    h1 { margin: 0; color: #112a23; font-size: 20pt; line-height: 1.1; font-weight: 800; }
    .hero-meta { display: flex; flex-wrap: wrap; gap: 1.5mm; align-items: center; margin-top: 2mm; color: #66736d; font-size: 8.3pt; }
    .metrics { display: grid; grid-template-columns: repeat(3,1fr); margin: 3.5mm 0 1mm; overflow: hidden; border: .35mm solid #d9e6df; border-radius: 3mm; background: #edf7f1; break-inside: avoid; }
    .metric { padding: 2.6mm; text-align: center; border-right: .3mm solid #d9e6df; }.metric:last-child{border:0}.metric strong{display:block;color:#0f4f36;font-size:15pt}.metric span{color:#66736d;font-size:7pt;font-weight:650}
    section { margin-top: 3.5mm; } section > h2 { margin: 0 0 1.8mm; padding-bottom: 1.1mm; border-bottom: .5mm solid #197a52; color: #0f4f36; font-size: 13.5pt; line-height: 1.18; font-weight: 800; break-after: avoid; }
    section.compact { break-inside: avoid; }
    h3 { margin: 2.5mm 0 1mm; color: #112a23; font-size: 10.5pt; font-weight: 800; } p { margin: 0 0 1.6mm; } ul { margin: .7mm 0 0; padding-inline-start: 6mm; } li { margin-bottom: 1.1mm; }
    .signals { display: grid; grid-template-columns: 1fr 1fr; gap: 2.5mm; }.signal { padding: 2.8mm; border: .3mm solid #d9e6df; border-radius: 2.5mm; background: #f8fbf9; break-inside: avoid; }.signal:only-child{grid-column:1/-1}.signal-heading{display:flex;justify-content:space-between;gap:3mm;color:#0f4f36;text-transform:capitalize}.signal-heading span{color:#197a52;font-weight:800}.signal p{margin:1mm 0;font-weight:700}.signal blockquote{margin:0;padding-inline-start:2.5mm;border-inline-start:.6mm solid #c7ddd1;color:#66736d;font-size:7.5pt}
    .style-grid { display:grid;grid-template-columns:repeat(3,1fr);gap:0;overflow:hidden;border:.3mm solid #d9e6df;border-radius:2.5mm }.style-grid div{padding:3mm;border-right:.3mm solid #d9e6df}.style-grid div:last-child{border:0}.style-grid small{display:block;color:#66736d;font-size:7pt;font-weight:750;text-transform:uppercase}.style-grid strong{display:block;margin-top:1mm;color:#112a23}
    .commitment { display:grid;grid-template-columns:25mm minmax(0,1fr);gap:2mm;align-items:start }.commitment strong{color:#0f4f36}.note { margin-top: 3.5mm; padding: 2.8mm; border: .3mm solid #d9e6df; border-radius: 2.5mm; color: #52615a; background: #f8fbf9; font-size: 7.8pt; break-inside: avoid; }.note strong{color:#112a23;font-size:8.7pt}.audit-line{display:flex;flex-wrap:wrap;gap:1mm 2mm;align-items:baseline}.audit-line bdi{font-weight:700;color:#33473f}
  </style></head><body>
    <header><img src="${logoUrl}" alt=""><div><strong>AmirOS</strong><small>Private relationship intelligence</small></div></header>
    <main>
      <div class="hero"><div class="avatar">${profileImageUrl ? `<img src="${profileImageUrl}" alt="">` : escapeHtml(initials)}</div><div><p class="eyebrow">${input.isGroup ? "Group relationship profile" : "Private contact profile"}</p>${text(contactName,"h1")}<div class="hero-meta"><strong dir="auto">${escapeHtml(input.contact.relationship || "Contact")}</strong><span>•</span><strong dir="auto">${escapeHtml(input.contact.tone || "Automatic")}</strong><span>•</span><span dir="auto">${escapeHtml(input.contact.language || "Automatic")}</span></div></div></div>
      <div class="metrics"><div class="metric"><strong>${input.profile.sourceMessageCount}</strong><span>TRACKED MESSAGES</span></div><div class="metric"><strong>${insights.length}</strong><span>ACTIVE INSIGHTS</span></div><div class="metric"><strong>${commitments.length}</strong><span>OPEN COMMITMENTS</span></div></div>
      <section><h2>Profile analysis</h2>${summaryHtml(input.profile.summary)}</section>
      ${insights.length ? `<section><h2>Relationship intelligence</h2><div class="signals">${insightHtml}</div></section>` : ""}
      ${commitments.length ? `<section><h2>Open commitments</h2><ul>${commitmentHtml}</ul></section>` : ""}
      ${style ? `<section><h2>Communication style</h2>${text(style.summary,"p")}<div class="style-grid"><div><small>Message length</small>${text(style.messageLength,"strong")}</div><div><small>Emoji use</small>${text(style.emojiUse,"strong")}</div><div><small>Formality</small>${text(style.formality,"strong")}</div></div><ul>${style.replyGuidance.map((item) => `<li dir="auto">${escapeHtml(item)}</li>`).join("")}</ul></section>` : ""}
      ${group ? `<section><h2>Group dynamics</h2>${text(group.summary,"p")}${[["Decisions",group.decisions],["Tasks",group.tasks],["Open questions",group.unansweredQuestions]].map(([heading,values]) => (values as string[]).length ? `${text(heading,"h3")}<ul>${(values as string[]).map((item) => `<li dir="auto">${escapeHtml(item)}</li>`).join("")}</ul>` : "").join("")}</section>` : ""}
      ${input.manualMemory.length ? `<section class="compact"><h2>Operator-saved memory</h2><ul>${memoryHtml}</ul></section>` : ""}
      <div class="note"><strong>Analysis details</strong><div class="audit-line"><span>Generated</span><bdi dir="auto">${escapeHtml(generatedLabel)}</bdi><span>· ${input.profile.sourceMessageCount} tracked messages</span><span>· ${input.manualMemory.length} manual memory items</span></div>Inferences are evidence-backed where available, but may still be incomplete and should be reviewed.</div>
    </main>
    <footer><span>Private analysis generated locally by AmirOS</span><bdi dir="auto">${escapeHtml(generatedLabel)}</bdi></footer>
  </body></html>`;
}

export function safePdfFilename(contactName: string): string {
  const slug = contactName.normalize("NFKD").replace(/[^a-zA-Z0-9]+/g, "-").replace(/^-+|-+$/g, "").toLowerCase().slice(0, 60) || "contact";
  return `AmirOS-profile-${slug}.pdf`;
}

export async function generateContactProfilePdf(input: ContactProfilePdfInput): Promise<Buffer> {
  const logoUrl = await imageDataUrl(resolve("ui/public/amiros-mark-v2-cropped.png"));
  const require = createRequire(import.meta.url);
  const whatsappEntry = require.resolve("whatsapp-web.js");
  const puppeteerEntry = require.resolve("puppeteer", { paths: [dirname(whatsappEntry)] });
  const puppeteerModule = await import(pathToFileURL(puppeteerEntry).href) as {
    default: {
      launch(options: { headless: boolean; executablePath: string; args: string[] }): Promise<{
        newPage(): Promise<{
          setContent(html: string, options: { waitUntil: string }): Promise<void>;
          pdf(options: { format: string; printBackground: boolean; preferCSSPageSize: boolean }): Promise<Uint8Array>;
        }>;
        close(): Promise<void>;
      }>;
    };
  };
  const browser = await puppeteerModule.default.launch({
    headless: true,
    executablePath: chromeExecutable(),
    args: ["--no-sandbox", "--disable-setuid-sandbox", "--disable-background-networking"],
  });
  try {
    const page = await browser.newPage();
    await page.setContent(renderPdfHtml(input, logoUrl), { waitUntil: "load" });
    const pdf = Buffer.from(await page.pdf({ format: "A4", printBackground: true, preferCSSPageSize: true }));
    if (!pdf.subarray(0, 5).equals(Buffer.from("%PDF-"))) throw new Error("Profile PDF generation returned an invalid document");
    return pdf;
  } finally {
    await browser.close();
  }
}
